src/agent/coverage/src/code.rs

```diff
@@ -0,0 +1,331 @@
+                let sym_section_offset = sym_va - section_va;
+
+                // We have the file offset for the section via `sh_offset`, and the offset of the
+                // symbol within the section. From this, calculate the file offset for the symbol,
+                // which we can use to index into `data`.
+                let sym_file_offset = section.sh_offset + sym_section_offset;
+
+                let entry = Symbol {
+                    name,
+                    file_offset: sym_file_offset,
+                    image_offset,
+                    size: sym.st_size,
+                };
+
+                let inserted = symbols.index.insert(entry.clone());
+                if !inserted {
+                    log::error!("failed to insert symbol index entry: {:x?}", entry);
+                }
+            }
+        }
+
+        Ok(Self {
+            path,
+            base_va,
+            symbols,
+        })
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+#[serde(transparent)]
+pub struct SymbolIndex {
+    pub index: RegionIndex<Symbol>,
+}
+
+impl SymbolIndex {
+    pub fn iter(&self) -> impl Iterator<Item = &Symbol> {
+        self.index.iter()
+    }
+
+    /// Find the symbol metadata for the image-relative `offset`.
+    pub fn find(&self, offset: u64) -> Option<&Symbol> {
+        self.index.find(offset)
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct Symbol {
+    /// Raw symbol name, possibly mangled.
+    pub name: String,
+
+    /// File offset of the symbol definition in the on-disk module.
+    pub file_offset: u64,
+
+    /// Module-relative offset of the mapped symbol.
+    pub image_offset: u64,
+
+    /// Total size in bytes of the symbol definition.
+    pub size: u64,
+}
+
+impl Symbol {
+    pub fn contains_file_offset(&self, offset: u64) -> bool {
+        let lo = self.file_offset;
+        let hi = lo + self.size;
+        (lo..hi).contains(&offset)
+    }
+
+    pub fn contains_image_offset(&self, offset: u64) -> bool {
+        let lo = self.image_offset;
+        let hi = lo + self.size;
+        (lo..hi).contains(&offset)
+    }
+}
+
+/// Symbol metadata defines a `Region` relative to its process image.
+impl Region for Symbol {
+    fn base(&self) -> u64 {
+        self.image_offset
+    }
+
+    fn size(&self) -> u64 {
+        self.size
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct SymbolFilterSpec {
+    filters: HashMap<String, Filter>,
+}
+
+#[derive(Clone, Debug)]
+pub struct SymbolFilter {
+    /// Pre-compiled regex set for fast matching.
+    regexes: RegexSet,
+
+    /// Maps module name regexes to filters for symbol names.
+    filters: BTreeMap<usize, Filter>,
+}
+
+impl SymbolFilter {
+    pub fn new(mut spec: SymbolFilterSpec) -> Result<Self> {
```

Comment:
The construct of using `spec.filters.keys()` followed by the values later on by string/pattern is difficult to reason about.

This would be simplified using [RegexSet::patterns](https://docs.rs/regex/1.1.0/regex/struct.RegexSet.html), rather than saving a copy of the string representations of the regex.

Reply:
Sorry, I'm not really sure what you're suggesting here. I'll add some comments and we can revisit.

Reply:
Ah! Actually, there's some implementation-determined runtime behavior in the module matching that we really don't want. 

I'm going to revisit this to be array-based (on the user/JSON side), and I think that will make the implementation easier to read and understand.

Reply:
(I plan on addressing this in a follow-up PR).